ios/RCTWebRTC/WebRTCModule+Daily.m

```diff
@@ -144,58 +145,29 @@ - (void)audioSession:(RTCAudioSession *)audioSession didSetActive:(BOOL)active {
   // active. Otherwise, it will be applied when the session becomes active.
   RTCAudioSession *audioSession = RTCAudioSession.sharedInstance;
+  NSLog(@"[Daily] setDailyAudioMode: %@", audioMode);
   if (audioSession.isActive) {
     [self applyAudioMode:audioMode toSession:audioSession];
   }
 }
 
 - (void)applyAudioMode:(NSString *)audioMode toSession:(RTCAudioSession *)audioSession {
+  NSLog(@"[Daily] applyAudioMode: %@", audioMode);
   // Do nothing if we're attempting to "unset" the in-call audio mode (for now
   // it doesn't seem like there's anything to do).
   if ([audioMode isEqualToString:AUDIO_MODE_IDLE]) {
     return;
   }
   
-  // Ducking other apps' audio implicitly enables allowing mixing audio with
-  // other apps, which allows this app to stay alive in the backgrounnd during
-  // a call (assuming it has the voip background mode set).
-  AVAudioSessionCategoryOptions categoryOptions = (AVAudioSessionCategoryOptionAllowBluetooth |
-                                                   AVAudioSessionCategoryOptionDuckOthers);
-  if ([audioMode isEqualToString:AUDIO_MODE_VIDEO_CALL]) {
-    categoryOptions |= AVAudioSessionCategoryOptionDefaultToSpeaker;
-  }
-  [self audioSessionSetCategory:AVAudioSessionCategoryPlayAndRecord toSession:audioSession options:categoryOptions];
-  
-  
-  NSString *mode = ([audioMode isEqualToString:AUDIO_MODE_VIDEO_CALL] ?
-                    AVAudioSessionModeVideoChat :
-                    AVAudioSessionModeVoiceChat);
-  [self audioSessionSetMode:mode toSession:audioSession];
-}
-
-- (void)audioSessionSetCategory:(NSString *)audioCategory
-                      toSession:(RTCAudioSession *)audioSession
-                        options:(AVAudioSessionCategoryOptions)options
-{
-  @try {
-    [audioSession setCategory:audioCategory
-                  withOptions:options
-                        error:nil];
-    NSLog(@"[Daily] audioSession.setCategory: %@, withOptions: %lu success", audioCategory, (unsigned long)options);
-  } @catch (NSException *e) {
-    NSLog(@"[Daily] audioSession.setCategory: %@, withOptions: %lu fail: %@", audioCategory, (unsigned long)options, e.reason);
-  }
-}
-
-- (void)audioSessionSetMode:(NSString *)audioMode
-                  toSession:(RTCAudioSession *)audioSession
-{
-  @try {
-    [audioSession setMode:audioMode error:nil];
-    NSLog(@"[Daily] audioSession.setMode(%@) success", audioMode);
-  } @catch (NSException *e) {
-    NSLog(@"[Daily] audioSession.setMode(%@) fail: %@", audioMode, e.reason);
+  NSString * audioDevice = [audioMode isEqualToString:AUDIO_MODE_VIDEO_CALL] ? SPEAKERPHONE_DEVICE_ID: WIRED_OR_EARPIECE_DEVICE_ID;
+  // override the default audio device in case we have a wired or bluetooth headset
+  if(self.hasWiredHeadset){
+      audioDevice = WIRED_OR_EARPIECE_DEVICE_ID;
+  }else if(self.hasBluetoothDevice){
+      audioDevice = BLUETOOTH_DEVICE_ID;
   }
+    
+  [self setAudioDevice:audioDevice];
```

Comment:
issue here (i _think_, haven't tested): we're _first_ mapping from desired mode (video, voice) to desired device, and then applying the device.

that means that if, say, the user asked to be in video call mode and started the call with a headset plugged in, it wouldn't automatically go to speakerphone when the headset was unplugged.

what i think this logic should look like:

```
if ([audioMode isEqualToString:AUDIO_MODE_VIDEO_CALL]) {
  [self setAudioMode:AUDIO_MODE_VIDEO_CALL];
}
else if ([audioMode isEqualToString:AUDIO_MODE_VIDEO_CALL]) {
  [self setAudioMode:AUDIO_MODE_VIDEO_CALL];
}
else { // AUDIO_MODE_USER_SPECIFIED
  NSString * audioDevice = // ...look up what device the user last asked for
  [self setAudioDevice:audioDevice];
}
```

Reply:
Fixed. Take a look on my last commit: _Refactor to keep the user selected device, and don't change the audio mode unless the user has explicitly changed the device._